File: credit.c
#include <cs50.h>
#include <stdio.h>

string check_type(long n);
string check_validity(long n, string c);

int main(void)
{
    // get card number from user
    long card_number = get_long("Number: ");

    // check for possible card type
    string card_type = check_type(card_number);

    // check for card validity
    card_type = check_validity(card_number, card_type);

    printf("%s\n", card_type);
}



string check_type(long n)
{
    string s;
    // get last digits per card type
    long v = n / 1000000000000;
    long m = n / 100000000000000;
    long a = n / 10000000000000;
    // check for VISA 13- and 16-digit, start with 4
    if (v == 4 || v / 1000  == 4)
    {
        s = "VISA";
    }
    // check for MASTERCARD 16-digit, start with 51, 52, 53, 54, or 55
    else if (m == 51 || m == 52 || m == 53 || m == 54 || m == 55)
    {
        s = "MASTERCARD";
    }
    // check for AMERICAN EXPRESS 15-digit, start with 34 or 37
    else if (a == 34 || a == 37)
    {
        s = "AMEX";
    }
    else
    {
        s = "INVALID";
    }
    return s;
}



string check_validity(long n, string c)
{
    long d; // variable for each of the digits
    int sum;
    int t = 1;
    for (int i = 0; i < 16; i++)
    {
        d = (n / t) % 10;
        if (i % 2 == 1) // checks if i is an odd number
        {
            sum = sum + n;
        }
        else // if i is even we multiply by 2
        {
            sum = sum + (2 * n);
        }
        t = t * 10; // makes sure that we get the next digit for the next iteration
    }

    if (sum % 10 != 0)
    {
        c = "INVALID";
    }
    return c;
}

File: credit.py
import cs50
import re
import math


def main():

    card_number = cs50.get_int("Number: ")
    card_type = check_type(str(card_number))

    if not valid(card_number):
        card_type = "INVALID"

    print(card_type)


def check_type(card_number):
    # RegEx of the 3 different card types
    amex = re.match('34\d{13,13}|37\d{13,13}', card_number)
    mastercard = re.match('5[1-5]\d{14,14}', card_number)
    visa = re.match('4\d{12,12}|4\d{15,15}', card_number)

    if amex:
        card_type = "AMEX"
    elif mastercard:
        card_type = "MASTERCARD"
    elif visa:
        card_type = "VISA"
    else:
        card_type = "INVALID"

    return card_type


def valid(card_number):
    # Luhn's algorithm
    sum = 0
    for i in range(int(math.log10(card_number) + 1)):
        # get the last digit and multiple by 2 every other digit
        d = card_number % 10
        if i % 2 == 0:
            sum += d
        else:
            if (d * 2) < 10:
                sum += (d * 2)
            else:
                sum += ((d * 2) % 10) + 1
        # iterate to the second to the last digit
        card_number = int(card_number / 10)

    if sum % 10 == 0:
        return True
    else:
        return False

if __name__ == "__main__":
    main()
